File: my_av.py
#!/usr/bin/python3
import re


def url_euristic_1(url_local, domains_local):
    """Rezolva primul euristic"""
    check = 0
    for line in domains_local:
        domain = line.strip()
        if domain in url_local:
            check = 1
            break
    domains_local.seek(0)
    return check


def url_euristic_2(url_local):
    """Rezolva al doilea euristic"""
    bad = ".exe"
    if bad in url_local:
        check = 1
    else:
        check = 0
    return check


def nr_digits(url_local):
    """Gaseste numarul de cifre dintr-un url/domeniu"""
    contor = 0
    for char in url_local:
        if char.isdigit():
            contor += 1
    return contor


def url_euristic_3(url_local):
    """Rezolva al treilea euristic"""
    domain = url_local.split('/')[0]
    if nr_digits(domain) > (0.1 * len(domain)):
        check = 1
    else:
        check = 0
    return check


def url_euristic_4(url_local):
    """Rezolva al patrulea euristic (gasit de mine)"""
    domain = url_local.split('/')[0]
    bad = "www"
    if bad in domain:
        check = 1
    else:
        check = 0
    return check

# ***************Functii pentru trafic malitios *******************


def get_index_of_parameter(parametru, antet_local):
    """Gaseste indexul unui parametru in antetul din traffic_in"""
    string = antet_local.split(",")
    for index, categorie in enumerate(string, 1):
        if parametru in categorie:
            return index
    return None


def traffic_euristic_1(traffic_local, antet_local):
    """Rezolva primul euristic"""
    parametru = "flow_duration"
    contor = get_index_of_parameter(parametru, antet_local)

    flow_duration = traffic_local.split(',')[contor-1]
    flow_duration = re.split(r'[ :]', flow_duration)

    days = int(flow_duration[0])
    ore = int(flow_duration[2])
    minute = int(flow_duration[3])
    sec = float(flow_duration[4])

    if days > 0 or ore > 0 or minute > 0 or sec > 1:
        check = 1
    else:
        check = 0
    return check


def traffic_euristic_2(traffic_local, antet_local):
    """Rezolva al doilea euristic"""
    parametru = "flow_pkts_payload.avg"
    contor = get_index_of_parameter(parametru, antet_local)
    flow_pkts_payload = traffic_local.split(',')[contor-1]
    if float(flow_pkts_payload) != 0:
        return 1
    return 0

# ***************Incepe "main-ul"*******************


# ***************Url malitios*******************
with open('./data/urls/urls.in', 'r', encoding='utf-8') as url_in, \
        open('urls-predictions.out', 'w', encoding='utf-8') as url_out, \
        open('./data/urls/domains_database', 'r', encoding='utf-8') as domains_in:

    for url in url_in:
        URL_E1 = url_euristic_1(url, domains_in)
        URL_E2 = url_euristic_2(url)
        URL_E3 = url_euristic_3(url)
        URL_E4 = url_euristic_4(url)

        if URL_E1 == 1 or URL_E2 == 1 or URL_E3 == 1 or URL_E4 == 1:
            url_out.write("1\n")
        else:
            url_out.write("0\n")

# ***************Trafic malitios*******************

with open('./data/traffic/traffic.in', 'r', encoding='utf-8') as traffic_in, \
        open('traffic-predictions.out', 'w', encoding='utf-8') as traffic_out:

    antet = traffic_in.readline()
    for traffic in traffic_in:
        TRAFFIC_E1 = traffic_euristic_1(traffic, antet)
        TRAFFIC_E2 = traffic_euristic_2(traffic, antet)
        if TRAFFIC_E1 == 1 and TRAFFIC_E2 == 1:
            traffic_out.write("1\n")
        else:
            traffic_out.write("0\n")

File: my_av.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#define max1 1000
#define max2 300

// ***************Functii pentru detectarea url-urilor malitioase *******************
int url_euristic_1(char *url, FILE *domains_in)
{
	char *domain = (char *)calloc(max1, sizeof(char));
	int ok = 0;
	while (fgets(domain, max1, domains_in)) {
		domain[strlen(domain) - 1] = '\0';
		if (strstr(url, domain)) {
			ok = 1;
			break;
		}
	}
	free(domain);
	rewind(domains_in);
	return ok;
}

int url_euristic_2(char *url)
{
	char *bad = ".exe";
	if (strstr(url, bad))
		return 1;
	return 0;
}

int nr_digits(char *url)
{
	char *numbers = "0123456789";
	int count = 0;
	for (int i = 0; i < strlen(numbers); i++) {
		if (strchr(url, numbers[i])) {
			char *p = strchr(url, numbers[i]);
			count++;
			while (p) {
				p = strchr(p + 1, numbers[i]);
				count++;
			}
		}
	}

	return count;
}

int url_euristic_3(char *url)
{
	char *domain = strtok(url, "/");
	if (nr_digits(domain) > (0.1 * strlen(domain)))
		return 1;
	return 0;
}

int url_euristic_4(char *url)
{
	char *domain = strtok(url, "/");
	if (strstr(domain, "www"))
		return 1;
	return 0;
}

// ***************Functii pentru detectarea traficului suspect *******************

int get_index_of_parameter(char *parametru, char *antet)
{
	char *p = strtok(antet, ",");
	int contor = 1;
	while (p) {
		if (strlen(p) < strlen(parametru)) {
			if (strstr(parametru, p))
				break;
		} else {
			if (strstr(p, parametru))
				break;
		}
		p = strtok(NULL, ",");
		contor++;
	}
	return contor;
}

int traffic_euristic_1(char *traffic, char *antet)
{
	char *parametru = "flow_duration";
	int contor = get_index_of_parameter(parametru, antet);

	char *flow_duration = strtok(traffic, ",");
	for (int i = 1; i < contor; i++)
		flow_duration = strtok(NULL, ",");

	int days = 0;
	if (sscanf(strtok(flow_duration, " :"), "%d", &days) == 1)
		; // do nothing

	strtok(NULL, " :"); // cuvantul "days"

	int h = 0;
	if (sscanf(strtok(NULL, " :"), "%d", &h) == 1) // ore
		; // do nothing

	int min = 0;
	if (sscanf(strtok(NULL, " :"), "%d", &min) == 1) // minute
		; // do nothing

	float sec = 0;
	if (sscanf(strtok(NULL, " :"), "%f", &sec) == 1) // secunde
		; // do nothing
	if (days > 0 || h > 0 || min > 0 || sec > 1)
		return 1;

	return 0;
}

int traffic_euristic_2(char *traffic, char *antet)
{
	char *parametru = "flow_pkts_payload.avg";
	int contor = get_index_of_parameter(parametru, antet);
	char *flow_pkts_payload = strtok(traffic, ",");

	for (int i = 1; i < contor; i++)
		flow_pkts_payload = strtok(NULL, ",");

	float value = 0;
	if (sscanf(flow_pkts_payload, "%f", &value) == 1)
		; // do nothing
	if (value != 0)
		return 1;
	return 0;
}

int main(void)
{
	// ***************Url malitios*******************
	char *file_urls = "./data/urls/urls.in";
	char *file_domains = "./data/urls/domains_database";

	FILE *url_in = fopen(file_urls, "r");
	FILE *url_out = fopen("urls-predictions.out", "w");
	FILE *domains_in = fopen(file_domains, "r");

	int url_e1 = 0, url_e2 = 0, url_e3 = 0, url_e4 = 0;
	int len = 0;

	char *url = (char *)calloc(max1, sizeof(char));

	while (fgets(url, max1, url_in)) {
		url_e1 = url_euristic_1(url, domains_in);
		url_e2 = url_euristic_2(url);
		url_e3 = url_euristic_3(url);
		url_e4 = url_euristic_4(url);
		if (url_e1 == 1 || url_e2 == 1 || url_e3 == 1 || url_e4 == 1)
			fprintf(url_out, "1\n");
		else
			fprintf(url_out, "0\n");
	}

	fclose(url_in); fclose(url_out); fclose(domains_in);
	free(url);

	// ***************Trafic malitios*******************
	char *file_traffics = "./data/traffic/traffic.in";

	FILE *traffic_in = fopen(file_traffics, "r");
	FILE *traffic_out = fopen("traffic-predictions.out", "w");

	int traffic_e1 = 0, traffic_e2 = 0;

	char *traffic = (char *)calloc(max2, sizeof(char));
	char *antet = (char *)calloc(max2, sizeof(char));
	char *aux1 = (char *)calloc(max2, sizeof(char));
	char *aux2 = (char *)calloc(max2, sizeof(char));

	fgets(antet, max2, traffic_in);

	while (fgets(traffic, max2, traffic_in)) {
		strcpy(aux1, traffic);
		strcpy(aux2, antet);
		traffic_e1 = traffic_euristic_1(aux1, aux2);
		strcpy(aux1, traffic);
		strcpy(aux2, antet);
		traffic_e2 = traffic_euristic_2(aux1, aux2);
		if (traffic_e1 == 1 && traffic_e2 == 1)
			fprintf(traffic_out, "1\n");
		else
			fprintf(traffic_out, "0\n");
	}

	fclose(traffic_in); fclose(traffic_out);
	free(traffic);
	free(antet);
	free(aux1);
	free(aux2);
	return 0;
}
